### ISSUE TITLE
increase max lsig program size

### ISSUE THREAD

**Issue by joe-p** [algorand/go-algorand#5990]
I'm creating this issue to continue the discussion on lsig size raised in #5943 by @giuliop. Below is his original post

> We keep of course a limit on the program size for lsig and apps. The apps one is fine as discussed above.
The lsig one feels very limited. For instance, there is a limit of 2048 bytes on app call arguments and a limit of 1024 bytes on note fields, both larger than the 1000 byte limit on lsigs.
I think it would be more balanced to set the lsig program limit to 4096 bytes (half the max for an app teal size), or at least 2048 bytes, as one page of app program.
In the future we can optimize how we store those transactions in the ledger by caching the lsigs and only storing their hash if we don't do that already. I guess we could also optimize how we broadcast them by sending their hash only once they are 'cached' by the network.
This, and the P2P gossip network, are to me the most important changes to the AVM right now.
The latter because decentralization is the only reason we are all here after all and this one for the developer experience.

I personally think 4k is reasonable considering with two transactions (one lsig + one appl) you can get access to > 10k bytes. The lsig program itself is 1k and then the second transaction can provide byte storage in note, args, and programs. 

We could also potentially "pool" the size together. Meaning I can have a group with one lsig that is 15k with 15 other lsigs that are only a couple of bytes each.

**Comment by giuliop** [comment]
👍

I think the easiest thing is to simply adjust the lsig teal max program size to 4k which is more in line with the other limits of the AVM.

The other alternative is to go down the rabbit hole of having variable transaction fees that scale with transaction size and enforce only that transactions (or transaction groups) fit in a block which I would resist as long as possible to keep things simple

**Comment by jannotti** [comment]
I do not think this is a good idea, since it makes it easy to bloat transaction sizes.  This is already a vector to annoy the network. Thankfully nobody has bothered yet, but it would be annoying for someone to fill blocks with cheap transactions that max out every possible field.  Adding 3k to each transaction would be only help them out.

I claim that lsigs can and should be short.  The main remaining use for lsigs is their high compute budget.  From what I understand, that's why they're being brought up here - to use the expensive elliptic curve opcodes that enable checking zk proofs.  Suppose, for example, that you need to use `ec_pairing_check` - it costs many thousands in opcode budget, so you surely want to do it in lsigs. You might try to put related code in the lsig, so you might run up against 1k length quickly.  I would propose that instead you separate your on-chain code into two pieces, the part that figures out what needs to be fed to `ec_pairing_check`, and the logicsig that performs the pairing checks.   The app can't feed data to the lsig, of course, so instead, you determine  what the app will want checked offline, and put those inputs in the application args.  The app runs, and simply confirms that the args supplied match what it wants checked.  The lsig "dumbly" checks the application args, not really knowing why. So it has already confirmed that if the args match the app's intent, they pass `ec_pairing_check`.  It's a bit convoluted, as you're now doing parts of the work in three places: offline, app, and lsig. But it allows each thing to be done where it can be done well.

**Comment by jannotti** [comment]
I'm mostly convinced by this:

>But you can already send a transaction group with 16 lsigs of 1kb each for a total of 16kb.
If we are happy with this we should be happy with a transaction group with a single lsig of 4kb.

I think we should probably pool the total allowable lsig size.  I _don't_ think we should try to avoid the creation of "dummy" transactions by using fee accounting, but then I'm not convinced we should do it for apps either.  For those who do want to do it, and are unconvinced of the complexity, think about how the following things interact: You paid extra because you want your apps to run longer, you paid extra because you want your lsigs to run longer (or be bigger), you paid extra because of congestion, you paid extra to register for incentives.  I don't think we should try to untangle that in the short, or even medium, term.

So, with that in mind, when I say I'm "mostly" convinced by the argument I quoted, I _don't_ mean we should allow a big lsig in a group containing a single txn that pays a large fee.  But I'd be happy saying that the total number of lsig bytes in a group should be 1000*numTxnsInGroup.

**Comment by giuliop** [comment]
I'd be happy too if we can have the total number of lsig bytes in a group as be 1000*numTxnsInGroup.
It achieves the same end result as one big lsig in the end, with just bit of extra complexity pushed to the front-end in constructing transaction groups with dummy lsig calls.

As for the complexity of fee accounting, would you feel the same way if we had to implement everything from scratch instead of modifying the existing code? I am asking because I am struggling to see the complexity you see and was wondering if it's because I don't know the existing code and how complex is it to change or if it's because I have some logical gaps in my thinking.

**Comment by giuliop** [comment]
I've created a [PR](https://github.com/algorand/go-algorand/pull/6057) to implement pooling across a transaction group as suggested

**Comment by giuliop** [comment]
Closing as completed
